src/adapter/src/coord/timestamp_selection.rs

```diff
@@ -277,6 +277,124 @@ pub trait TimestampProvider {
         ),
         AdapterError,
     > {
+        use constraints::{Constraints, Preference, Reason};
+
+        // First, we acquire read holds that will ensure the queried collections
+        // stay queryable at the chosen timestamp.
+        let read_holds = self.acquire_read_holds(id_bundle);
+        let timeline = Self::get_timeline(timeline_context);
+
+        // To be filled by strong session serializable logic, should it be engaged.
+        let mut session_oracle_read_ts = None;
+
+        // We start by establishing the hard constraints that must be applied to timestamp determination.
+        // These constraints are derived from the input arguments, and properties of the collections involved.
+        // TODO: Many of the constraints are expressed obliquely, and could be made more direct.
+        let constraints = {
+            // Constraints we will populate through a sequence of opinions.
+            let mut constraints = Constraints::default();
+
+            // First, we have validity constraints from the `id_bundle` argument which indicates
+            // which collections we are reading from.
+            // TODO: Refine the detail about which identifiers are binding and which are not.
+            let since = self.least_valid_read(&read_holds);
+            let storage = id_bundle.storage_ids.iter().cloned().collect();
+            constraints
+                .lower
+                .push((since.clone(), Reason::StorageInput(storage)));
```

Comment:
The `since` here and for compute below is over the entire bundle, so could be over stating its needs. It perhaps makes sense to teach `ReadHoldsInner` about these separately so this constraint is more accurately expressed.